crates/goose/src/agents/agent.rs

```diff
@@ -0,0 +1,396 @@
+use anyhow::Result;
+use async_stream;
+use async_trait::async_trait;
+use futures::stream::BoxStream;
+use rust_decimal_macros::dec;
+use serde_json::json;
+use std::collections::HashMap;
+use tokio::sync::Mutex;
+
+use crate::errors::{AgentError, AgentResult};
+use crate::message::{Message, ToolRequest};
+use crate::prompt_template::load_prompt_file;
+use crate::providers::base::{Provider, ProviderUsage};
+use crate::systems::System;
+use crate::token_counter::TokenCounter;
+use mcp_core::{Content, Resource, Tool, ToolCall};
+use serde::Serialize;
+
+// used to sort resources by priority within error margin
+const PRIORITY_EPSILON: f32 = 0.001;
+
+#[derive(Clone, Debug, Serialize)]
+struct SystemInfo {
+    name: String,
+    description: String,
+    instructions: String,
+}
+
+impl SystemInfo {
+    fn new(name: &str, description: &str, instructions: &str) -> Self {
+        Self {
+            name: name.to_string(),
+            description: description.to_string(),
+            instructions: instructions.to_string(),
+        }
+    }
+}
+
+#[derive(Clone, Debug, Serialize)]
+struct SystemStatus {
+    name: String,
+    status: String,
+}
+
+impl SystemStatus {
+    fn new(name: &str, status: String) -> Self {
+        Self {
+            name: name.to_string(),
+            status,
+        }
+    }
+}
+
+/// Core trait defining the behavior of an Agent
+#[async_trait]
+pub trait Agent: Send + Sync {
```

Comment:
Refactored, but ended up adding a usage method as well since we use it in cli and was kind of awkward to add it on to the manager (which implements the logic).